Compiler/src/ssair/EscapeAnalysis.jl

```diff
@@ -39,27 +43,177 @@ end
 
 include("disjoint_set.jl")
 
-const AInfo = IdSet{Any}
+abstract type MemoryKind end
+struct UninitializedMemory <: MemoryKind end
+struct CallerMemory <: MemoryKind
+    id::Int
+    arg::Argument # TODO arg::Union{Argument,CallerMemory}
+    fidx::Int
+end
+# struct SSAValueMemory <: MemoryKind end
+# struct GlobalRefMemory <: MemoryKind end
+# struct LiteralMemory <: MemoryKind end
+
+# TODO [^CFG-aware `MemoryInfo`]:
+# By incorporating some form of CFG information into `MemoryInfo`, it becomes possible
+# to enable load-forwarding even in cases where conflicts occur by inserting φ-nodes.
+
+abstract type MemoryInfo end
+struct MustAliasMemoryInfo <: MemoryInfo
+    alias::Any # anything that is valid as IR elements (e.g. `SSAValue`, `Argument`, `GlobalRef`, literals, and `MemoryKind`)
+    MustAliasMemoryInfo(@nospecialize alias) = new(alias)
+end
+struct MayAliasMemoryInfo <: MemoryInfo
+    aliases::IdSet{Any}
+    function MayAliasMemoryInfo(aliases::IdSet{Any})
+        @assert length(aliases) ≥ 2
+        return new(aliases)
+    end
+end
+struct UnknownMemoryInfo <: MemoryInfo end # not part of the `⊑ₘ` lattice, just a marker for `ssamemoryinfo`
+const ⊤ₘ = UnknownMemoryInfo()
+
+x::MemoryInfo == y::MemoryInfo = begin
+    @nospecialize
+    if x isa MayAliasMemoryInfo
+        return y isa MayAliasMemoryInfo && x.aliases == y.aliases
+    else
+        x = x::MustAliasMemoryInfo
+        y isa MustAliasMemoryInfo || return false
+        return x === y
+    end
+end
+function copy(x::MemoryInfo)
+    @nospecialize x
+    if x isa MayAliasMemoryInfo
+        return MayAliasMemoryInfo(copy(x.aliases))
+    end
+    return x
+end
+
+abstract type ObjectInfo end
+struct HasUnanalyzedMemory <: ObjectInfo end
```

Comment:
```suggestion
struct NoMemoryContents <: ObjectInfo end
```
Just a suggestion, but maybe closer to the intuition of this type as the oinfo lattice bottom